File: project2/src/memory.rs
use crate::cpu::RegisterFile;
use std::sync::{Arc, Mutex};

#[derive(Clone)]
pub struct Memory {
    pub size: usize,
    pub arr: Vec<i32>,
}

impl Memory {
    pub fn new(mem_size: usize) -> Memory {
        Self {
            size: mem_size,
            arr: vec![0; mem_size],
        }
    }
}

pub fn mem_init(mem_size: usize) -> Memory {
    Memory::new(mem_size)
}

pub fn mem_read(regs: &mut RegisterFile, mem_arc: &Arc<Mutex<Memory>>) {
    if regs.mar.reg_val < 0 {
        panic!("invalid mem address: {}. Exiting.", regs.mar.reg_val);
    }
    regs.mbr.reg_val = mem_arc.lock().unwrap().arr[regs.mar.reg_val as usize];
}

pub fn mem_write(regs: &RegisterFile, mem_arc: &Arc<Mutex<Memory>>) {
    if regs.mar.reg_val < 0 {
        panic!("invalid mem address: {}. Exiting. ", regs.mar.reg_val);
    }
    mem_arc.lock().unwrap().arr[regs.mar.reg_val as usize] = regs.mbr.reg_val;
}

pub fn mem_dump(mem: &Memory) {
    println!("===========================================");
    println!("           Memory Dump: Size = {}", mem.size);
    println!("===========================================");
    println!("Address: Contents");

    for i in 0..mem.size {
        println!("{}: {}", i, mem.arr[i]);
    }
    print!("\n");
}

File: sim-os-c/project2/memory.c
#include <stdlib.h>
#include "computer.h"

Memory mem = {0, NULL};
extern Register_File regs;

void mem_init(unsigned int M)
{
    mem.size = M;
    mem.mem_arr = calloc(M, sizeof(unsigned int));
    for (int i = 0; i < M; ++i)
    {
        mem.mem_arr[i] = 0;
    }
}

void mem_read()
{
    regs.MBR.reg_val = mem.mem_arr[regs.MAR.reg_val];
}

void mem_write()
{
    mem.mem_arr[regs.MAR.reg_val] = regs.MBR.reg_val;
}

void mem_dump()
{
    printf("===========================================\n");
    printf("           Memory Dump: Size = %d\n", mem.size);
    printf("===========================================\n");
    printf("Address: Contents\n");
    for (int i = 0; i < mem.size; ++i)
    {
        printf("%d: %d\n", i, mem.mem_arr[i]);
    }
    printf("\n");
}